### PR TITLE
Feature/android diff script

### PR DESCRIPTION
**Pull request by guidotheelen**
Add the android diffscript to this codebase. Allowing to check the emoji fonts as well.

Steps to test it:
- Install this project (README)
- Clone the noto-emoji (https://github.com/googlefonts/noto-emoji) project (Or use other Emoji.ttf files)
- Run the following command in the root of this project:
python Lib/diffenator path/to/noto-emoji/api_level/25/NotoColorEmoji.ttf path/to/noto-emoji/api_level/29/NotoColorEmoji.ttf -r=./comparison_gifs
- Check the output and the gifs in comparison_gifs

### CODE REVIEW

File: Lib/diffenator/diff.py [hunk 1]
```diff
@@ -340,13 +417,124 @@ def diff_glyphs(font_before, font_after,
...
+            codepoints = [glyph_to_codepoint_map[x] for x in glyph_names]
+            codepoint_map[codepoint_to_string(codepoints)] = ligature.LigGlyph
+
+
+def read_cbdt(ttf):
+  cbdt = ttf['CBDT']
+  glyph_to_image = {}
+  for strike_data in cbdt.strikeData:
+    for key, data in strike_data.items():
+      data.decompile
```
Comment (obeleh) [inline Lib/diffenator/diff.py:515]: Is that a method call?

File: Lib/diffenator/diff.py [hunk 2]
```diff
@@ -340,13 +417,124 @@ def diff_glyphs(font_before, font_after,
...
+      glyph_to_image[key] = Image.open(io.BytesIO(data.imageData))
+  return glyph_to_image
+
+
+def similar_img(img1, img2):
+    # return if images are the same with accepting some changes
+    if img1 is None and img2 is None: return True
+    if img1 is None or img2 is None: return False
+    if not img1.size == img2.size: return False
```
Comment (obeleh) [inline Lib/diffenator/diff.py:524]: I would write this as
```python
if img1.size != img2.size: return False
```

File: Lib/diffenator/diff.py [hunk 3]
```diff
@@ -340,13 +417,124 @@ def diff_glyphs(font_before, font_after,
...
+  cbdt = ttf['CBDT']
+  glyph_to_image = {}
+  for strike_data in cbdt.strikeData:
+    for key, data in strike_data.items():
+      data.decompile
+      glyph_to_image[key] = Image.open(io.BytesIO(data.imageData))
+  return glyph_to_image
+
+
+def similar_img(img1, img2):
```
Comment (obeleh) [inline Lib/diffenator/diff.py:520]: In a different project you used numpy to take the diff of an image. I would Imagine that that is a lot faster than running python code (numpy does it in C)

File: Lib/diffenator/diff.py [hunk 4]
```diff
@@ -773,4 +961,3 @@ def _modified_marks(marks_before, marks_after, thresh=4,
                 mark.pop(pos)
             table.append(mark)
     return table
-
```
Comment (obeleh) [inline Lib/diffenator/diff.py:776]: In python it is best practice to add an empty line on the end of the file ;)

File: Lib/diffenator/__main__.py [hunk 5]
```diff
@@ -103,26 +104,36 @@ def main():
...
-        print(diff.to_txt(args.output_lines))
+        if args.markdown:
+            print(diff.to_md(args.output_lines))
+        elif args.html:
+            print(diff.to_html(args.output_lines, image_dir=args.render_path))
+        else:
+            print(diff.to_txt(args.output_lines))
 
+    except Exception as ex:
+        if "Font contains no outlines" == str(ex):
```
[the remark this answers is not in the document]
Comment (guidotheelen) [reply Lib/diffenator/__main__.py:129]: @RoelN Is there an easy way to detect if a font has outlines?
When comparing two emoji fonts we get the following error: `Font contains no outlines`

File: Lib/diffenator/__main__.py [hunk 6]
```diff
@@ -89,7 +90,7 @@ def main():
     parser.add_argument('-rd', '--render_diffs', action='store_true',
                         help=("Render glyphs with hb-view and compare "
                               "pixel diffs."))
-    parser.add_argument('-r', '--render-path',
+    parser.add_argument('-r', '--render_path',
```
Comment (obeleh) [inline Lib/diffenator/__main__.py:93]: Was this intentional? It is more CLI-like to have hyphens in your parameters. Argparse will convert hyphens into underscores. See the [docs](https://docs.python.org/dev/library/argparse.html#dest)

File: Lib/diffenator/diff.py [hunk 7]
```diff
@@ -287,6 +292,78 @@ def _modified_names(names_before, names_after):
...
+        font1_cbdt = read_cbdt(ttf)
+        read_cmap12(ttf, glyph_to_codepoint_map_1, codepoint_map_1)
+        read_gsub(ttf, glyph_to_codepoint_map_1, codepoint_map_1)
+
+    with contextlib.closing(ttLib.TTFont(font_after)) as ttf:
+        font2_cbdt = read_cbdt(ttf)
+        read_cmap12(ttf, glyph_to_codepoint_map_2, codepoint_map_2)
+        read_gsub(ttf, glyph_to_codepoint_map_2, codepoint_map_2)
+
+    glyphs1 = set(font1_cbdt.keys())
```
[the remark this answers is not in the document]
Comment (guidotheelen) [reply Lib/diffenator/diff.py:312]: You are right, I don't 👍

File: Lib/diffenator/diff.py [hunk 8]
```diff
@@ -287,6 +292,78 @@ def _modified_names(names_before, names_after):
...
 
 
+def cbdt_diff_script(font_before, font_after, diff_options,
+                     thresh=0.00, scale_upms=True, render_diffs=False):
+    codepoint_map_1 = {}
+    codepoint_map_2 = {}
+    glyph_to_codepoint_map_1 = {}
+    glyph_to_codepoint_map_2 = {}
+
+    with contextlib.closing(ttLib.TTFont(font_before)) as ttf:
```
[the remark this answers is not in the document]
Comment (guidotheelen) [reply Lib/diffenator/diff.py:302]: Yeah cool right 😝

File: Lib/diffenator/diff.py [hunk 9]
```diff
@@ -287,6 +292,78 @@ def _modified_names(names_before, names_after):
...
+
+    codepoints_set1 = set(codepoint_map_1.keys())
+    codepoints_set2 = set(codepoint_map_2.keys())
+
+    # if codepoints_set1 != codepoints_set2:
+    #     print(f"Codepoints set has changed: : {codepoints_set1 ^ codepoints_set2}")
+
+    modified = set()
+
+    all_codepoints = set(codepoint_map_1.keys()).union(codepoint_map_2.keys())
```
[the remark this answers is not in the document]
Comment (guidotheelen) [reply Lib/diffenator/diff.py:323]: That's cleaner 🤘

File: Lib/diffenator/diff.py [hunk 10]
```diff
@@ -287,6 +292,78 @@ def _modified_names(names_before, names_after):
...
+    for key in all_codepoints:
+        glyph1 = codepoint_map_1[key] if key in codepoint_map_1 else None
```
[the remark this answers is not in the document]
Comment (guidotheelen) [reply Lib/diffenator/diff.py:325]: Nice one 👍

File: Lib/diffenator/diff.py (hunk 2 above)
Comment (guidotheelen) [reply Lib/diffenator/diff.py:524]: Sounds better 🤔

File: Lib/diffenator/diff.py (hunk 3 above)
Comment (guidotheelen) [reply Lib/diffenator/diff.py:520]: Ahh good to know when we run into some performance issues!

File: Lib/diffenator/diff.py (hunk 4 above)
Comment (guidotheelen) [reply Lib/diffenator/diff.py:776]: Check ✅

File: Lib/diffenator/__main__.py (hunk 6 above)
Comment (guidotheelen) [reply Lib/diffenator/__main__.py:93]: Ahh cool. that's exactly what I needed to know 🙌

File: Lib/diffenator/diff.py (hunk 1 above)
Comment (guidotheelen) [reply Lib/diffenator/diff.py:515]: Did not do anything, removed it!

**Pull request closed without merging**